src/expr/src/table_function/generate_subscripts.rs

```diff
@@ -0,0 +1,148 @@
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], 0, false) AS s;
+/// ----
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], 1, true) AS s;
+/// ----
+/// 3
+/// 2
+/// 1
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], 1, false) AS s;
+/// ----
+/// 1
+/// 2
+/// 3
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY[ARRAY['foo'], ARRAY['bar'], ARRAY[null]], 0, false) AS s;
+/// ----
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY[ARRAY['foo'], ARRAY['bar'], ARRAY[null]], 1, false) AS s;
+/// ----
+/// 1
+/// 2
+/// 3
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY[ARRAY['foo'], ARRAY['bar'], ARRAY[null]], 2, false) AS s;
+/// ----
+/// 1
+/// ```
+#[function("generate_subscripts(list, int32, boolean) -> setof int32")]
+fn generate_subscripts_reverse(
+    array: ListRef<'_>,
+    dim: i32,
+    reverse: bool,
+) -> Result<impl Iterator<Item = i32>> {
+    let (mut cur, end) = generate_subscripts_inner(array, dim, reverse);
+
+    // if we don't branch reverse inside, then we have to box it outside to have the same type
+    // different closures different anonymous types, or range and reverse range.
+    let next = move || {
+        if reverse {
+            if cur <= end {
+                return None;
+            }
+        } else if cur >= end {
+            return None;
+        }
+        let ret = cur;
+        cur += if reverse { -1 } else { 1 };
+        Some(ret)
+    };
+    Ok(std::iter::from_fn(next))
+}
+
+/// ```slt
+/// query I
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], -1, false) AS s;
+/// ----
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], 0, false) AS s;
+/// ----
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], 1) AS s;
+/// ----
+/// 1
+/// 2
+/// 3
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY['foo', 'bar', null], 1) AS s;
+/// ----
+/// 1
+/// 2
+/// 3
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY[ARRAY['foo'], ARRAY['bar'], ARRAY[null]], 0) AS s;
+/// ----
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY[ARRAY['foo'], ARRAY['bar'], ARRAY[null]], 1) AS s;
+/// ----
+/// 1
+/// 2
+/// 3
+///
+/// query I
+/// SELECT generate_subscripts(ARRAY[ARRAY['foo'], ARRAY['bar'], ARRAY[null]], 2) AS s;
+/// ----
+/// 1
+/// ```
+#[function("generate_subscripts(list, int32) -> setof int32")]
+fn generate_subscripts(array: ListRef<'_>, dim: i32) -> Result<impl Iterator<Item = i32>> {
+    generate_subscripts_reverse(array, dim, false)
+}
+
+fn generate_subscripts_inner(array: ListRef<'_>, dim: i32, reverse: bool) -> (i32, i32) {
+    let nothing = (0, 0);
+    match dim {
+        ..=0 => nothing,
+        1 => {
+            if reverse {
+                (array.len() as i32, 0)
+            } else {
+                (1, array.len() as i32 + 1)
+            }
+        }
+        // Although RW's array can be zig-zag, we just look at the first element.
```

Comment:
There are several alternatives:
* length of first element (this PR)
* max length of all elements
* length of any element when rectangular, or error when not
* always error on dim >= 2 (DuckDB and our current `array_length`)

We may keep the current behavior as experimental/undocumented, and make a breaking change of the zig-zag behavior later.

Reply:
thanks! let me create an issue to track this after the PR gets merged